packages/icon/package.py

```diff
@@ -723,3 +734,66 @@ def _compiler_is_mixed_gfortran(self):
 
         from spack.compilers.gcc import Gcc
         return fc_name in Gcc.fc_names
+
+    @property
+    def build_directory(self):
+        """Overrides function from spack.build_system.autotools
+        
+        By default build_directory is identical as configure_directory
+        To enable out-of-source builds this is not the case anymore
+        """
+
+        return self.stage.source_path
+
+    @property
+    def configure_directory(self):
+        """Returns the directory where 'configure' resides.
+
+        Overides function from spack.build_systems.autotools
+
+        CAUTION: Does only work if Spack is inside the git-repo
+                 of ICON, otherwise "git rev-pars --show-toplevel"
+                 fails!
+
+        """
+
+        Git = which('git', required=True)
+        git_root = Git('rev-parse', '--show-toplevel',
+                       output=str).replace("\n", "")
+        if git_root != self.stage.source_path:
+            # mark out-of-source build for function
+            # copy_runscript_related_input_files
+            self.out_of_source_build = True
+            return git_root
+        else:
+            return self.stage.source_path
+
+    def configure(self, spec, prefix):
+        if os.path.exists(os.path.join(self.build_directory, 'icon.mk')):
+            tty.warn('icon.mk already present -> skip configure stage',
+                     '\t delete "icon.mk to not skip configure')
+            return
```

Comment:
Wouldn't this cause
```
$ spack dev-build icon @develop gpu=none ~dace
$ spack dev-build icon @develop gpu=none +dace
```
to not reevaluate `configure_args` on the second command, therefore building again with `--disable-dace`?

Reply:
I know this is a violation of what Spack should do. For ICON we are in a grey-zone because it is also the package that is heavily used for development.

I wouldn't make this a general pattern for Autotools packages, most of them have a small configure script. But since ICON basically runs confiugure for 5-10 externals this takes 5+ mins. And for development this is annoying.

Reply:
You are right, in this case we end up with a wrong set of flags in the makefile.

Let me fix that.

Reply:
Are you aiming to advice users to do `-u build` to prevent an installation and to avoid that users have to first uninstall icon before running their next build?
Because this will result in just calling `make` and there's an easier way to do this. It's what MeteoSwiss does right now - from out internal docu about icon:

Subsequent builds can be started by loading the build environment and executing "make" in it. That's the [gnu make](https://www.gnu.org/software/make/), it has nothing to do with spack. It's just executed in a nested spack-build-env.

```
spack build-env icon @cpu -- make -j12
```

Reply:
To clarify:
The mechanism to skip configure is intended to work when using environments.